<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Case Study: Antibodies: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" type="text/css" href="assets/styles.css">
<script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="manifest" href="site.webmanifest">
<link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav l2d"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="l2d" src="assets/images/l2d-logo.svg"><abbr class="badge badge-light" title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught." style="background-color: #FF4955; border-radius: 5px">
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link" style="color: #000">
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
            Pre-Alpha
          </a>
          <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/aio.html';">Instructor View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav l2d" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="l2d" src="assets/images/l2d-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Case Study: Antibodies
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Case Study: Antibodies
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Case Study: Antibodies
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress l2d">
    <div class="progress-bar l2d" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
        <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
                <i class="search-icon" data-feather="x" role="img"></i>
            </button>
            <div class="sidebar-inner">
                <div class="row mobile-row">
                    <div class="col">
                        <div class="sidenav-view-selector">
                            <div class="accordion accordion-flush" id="accordionFlush9">
                                <div class="accordion-item">
 
                                    <h2 class="accordion-header" id="flush-headingNine">
                                        <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                                            <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i>
                                            Learner View
                                        </button>
                                    </h2>
                                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                                        <div class="accordion-body">
                                            <a href="instructor/aio.html">Instructor View</a>
                                        </div>
                                    </div>
                                </div>
                                <!--/div.accordion-item-->
                            </div>
                            <!--/div.accordion-flush-->
                        </div>
                        <!--div.sidenav-view-selector -->
                    </div>
                    <!--/div.col -->
                    
                    <hr>
</div>
                <!--/div.mobile-row -->

                <div class="accordion accordion-flush" id="accordionFlush11">
                    <div class="accordion-item">
                        <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                            <h2 class="accordion-header chapters" id="flush-headingEleven">
                                EPISODES
                            </h2>
                        </button>
                        <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">
                            <div class="accordion-body">
<div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="Machine_Learning_Part1.html">1. Separating Mouse and Human Antibody Sequences using Protein Encoding and Machine Learning Classifiers</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->
</div>
                        </div>
                    </div>
                     
                    <hr class="half-width">
<div class="accordion accordion-flush resources" id="accordionFlush12">
                        <div class="accordion-item">
                            <h2 class="accordion-header" id="flush-headingTwelve">
                                <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                                    RESOURCES
                                </button>
                            </h2>
                            <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                                <div class="accordion-body">
                                    <ul>
<li>
                                            <a href="key-points.html">Key Points</a>
                                        </li>
                                        <li>
                                            <a href="reference.html#glossary">Glossary</a>
                                        </li>
                                        <li>
                                            <a href="profiles.html">Learner Profiles</a>
                                        </li>
                                         <li><a href="reference.html">Reference</a></li>
                                    </ul>
</div>
                            </div>
                        </div>
                    </div>
                    <hr class="half-width resources">
<a href="aio.html">See all in one page</a>
                    
                    <br><hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">
                        
                    </div>
                </div>
                <!-- /div.accordion -->
            </div>
            <!-- /div.sidebar-inner -->
        </nav>
</div>
    <!-- /div.sidebar -->
</div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-Machine_Learning_Part1"><p>Content from <a href="Machine_Learning_Part1.html">Separating Mouse and Human Antibody Sequences using Protein Encoding and Machine Learning Classifiers</a></p>
<hr>
<p>Last updated on 2024-04-04 |
        
        <a href="https://github.com/LearnToDiscover/case-study-antibodies/edit/main/episodes/Machine_Learning_Part1.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How may we create a machine learning classifier that can tell apart
two groups of proteins?</li>
<li>How may we check for overfitting in a trained machine learning
classifier?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand how protein sequences can become readable to machine
learning predictors</li>
<li>Check for overfitted data by testing with a totally naive
dataset</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="introduction"><h2 class="section-heading">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>If we want to generate a classifier that observes the differences
between two groups of protein seequences, then we need to extract
numerical information from our sequences. This is called encoding and
can be done through a variety of ways inlcuding residue level encoding
of each amino acid in your sequences with a 1x20 vector, representing
the possibility of 20 amino acids at each residue. This is called One
Hot Encoding, but often leads to a sparse dataset which is not suitable
for machine learning tasks, and each sequence must be spaced out so they
are of equal length. Instead, in this example we use the physiochemical
properties that may be calculated from the sequence as our numeric
information (ElAbd <em>et al.</em>, 2020).</p>
<p>Antibodies are made up of two heavy and two light chains, however,
the functional antigen-binding domains are Fv fragments at each fork of
the “Y” shape. These Fv fragments are where the VH domain of a heavy
chain and VL domain of a light chain interact and so much study has been
dedicated to these regions. An antibody record is considered “paired”
when both the VH and VL sequences of one antibody are known. This
knowledge was rare in the past and came from studying individual
antibodies, however, the advent of B-cell encapsulation and Next
Generation Sequencing now allowed researchers to sequence a repertoire
of paired antibodies (Rajan <em>et al.</em>, 2018).</p>
<p>In this exercise, we will use a sample of 1000 Human and 1000 Mouse
paired antibodies taken from the Observed Antibody Space (Olsen <em>et
al.</em>, 2022) and develop a machine learning classifier to separate
them. Firstly, we will use Kidera Factors (Nakai <em>et al.</em>, 1988)
and Propythia (Sequeira <em>et al.</em>, 2022) to generate our encodings
from an input of Fasta formatted sequences. Secondly, we will split
those encodings into training and test datasets for a selection of
machine learning classifiers and plot our results, comparing both
methods of encodings.</p>
</section><section id="references"><h2 class="section-heading">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<hr class="half-width">
<p>ElAbd, H., Bromberg, Y., Hoarfrost, A., Lenz, T., Franke, A., &amp;
Wendorff, M. (2020). Amino acid encoding for deep learning applications.
BMC Bioinformatics, 21(1), 235. <a href="https://doi.org/10.1186/s12859-020-03546-x" class="external-link uri">https://doi.org/10.1186/s12859-020-03546-x</a></p>
<p>Nakai, K., Kidera, A., &amp; Kanehisa, M. (1988). Cluster analysis of
amino acid indices for prediction of protein structure and function.
Protein Engineering, Design and Selection, 2(2), 93-100 <a href="https://doi.org/10.1093/protein/2.2.93" class="external-link uri">https://doi.org/10.1093/protein/2.2.93</a></p>
<p>Olsen, T. H., Boyles, F., &amp; Deane, C. M. (2022). Observed
Antibody Space: A diverse database of cleaned, annotated, and translated
unpaired and paired antibody sequences. Protein Science, 31(1), 141–146.
<a href="https://doi.org/https://doi.org/10.1002/pro.4205" class="external-link uri">https://doi.org/https://doi.org/10.1002/pro.4205</a></p>
<p>Rajan, S., Kierny, M. R., Mercer, A., Wu, J., Tovchigrechko, A., Wu,
H., Dall′Acqua, W. F., Xiao, X., &amp; Chowdhury, P. S. (2018).
Recombinant human B cell repertoires enable screening for rare,
specific, and natively paired antibodies. Communications Biology, 1(1),
5. <a href="https://doi.org/10.1038/s42003-017-0006-2" class="external-link uri">https://doi.org/10.1038/s42003-017-0006-2</a></p>
<p>Sequeira, A. M., Lousa, D., &amp; Rocha, M. (2022). ProPythia: A
Python package for protein classification based on machine and deep
learning. Neurocomputing, 484, 172–182. <a href="https://doi.org/https://doi.org/10.1016/j.neucom.2021.07.102" class="external-link uri">https://doi.org/https://doi.org/10.1016/j.neucom.2021.07.102</a></p>
</section><section id="imports-and-requirements"><h2 class="section-heading">Imports and Requirements<a class="anchor" aria-label="anchor" href="#imports-and-requirements"></a>
</h2>
<hr class="half-width">
<p>The antibody encoding method we will be using is the Propythia
program. Before starting we recommend installing it by copying and
pasting the following command in your terminal/command line:</p>
<p><code>pip install propythia</code></p>
<p>The machine learning models that we will use are imported from the
sklearn package. We also suggest installing this via the following
command:</p>
<p><code>pip install scikit-learn</code></p>
<p>As this is a large package, rather than importing the whole package,
it is better practise to just import the fuctions we require. As this
usually leaves with a lot of import statements, we have broken them down
what imports are required at what stage of the process.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co">##Get Encodings###</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="im">from</span> pandas <span class="im">import</span> read_csv</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co">##Data preparation###</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="im">from</span> sklearn.utils <span class="im">import</span> check_random_state, shuffle</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="im">from</span> sklearn.model_selection <span class="im">import</span> train_test_split</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="im">from</span> sklearn.model_selection <span class="im">import</span> StratifiedShuffleSplit</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="im">from</span> sklearn.preprocessing <span class="im">import</span> StandardScaler</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="co">##Machine Learning Models###</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="im">from</span> sklearn.neural_network <span class="im">import</span> MLPClassifier</span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="im">from</span> sklearn.neighbors <span class="im">import</span> KNeighborsClassifier</span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="im">from</span> sklearn.svm <span class="im">import</span> SVC</span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a><span class="im">from</span> sklearn.gaussian_process <span class="im">import</span> GaussianProcessClassifier</span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a><span class="im">from</span> sklearn.gaussian_process.kernels <span class="im">import</span> RBF</span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a><span class="im">from</span> sklearn.tree <span class="im">import</span> DecisionTreeClassifier</span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a><span class="im">from</span> sklearn.ensemble <span class="im">import</span> RandomForestClassifier, AdaBoostClassifier</span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a><span class="im">from</span> sklearn.naive_bayes <span class="im">import</span> GaussianNB</span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a><span class="im">from</span> sklearn.discriminant_analysis <span class="im">import</span> QuadraticDiscriminantAnalysis</span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a><span class="im">from</span> sklearn.mixture <span class="im">import</span> GaussianMixture</span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a><span class="im">from</span> sklearn.cluster <span class="im">import</span> KMeans</span>
<span id="cb1-25"><a href="#cb1-25" tabindex="-1"></a><span class="im">from</span> sklearn.cluster <span class="im">import</span> DBSCAN</span>
<span id="cb1-26"><a href="#cb1-26" tabindex="-1"></a><span class="im">from</span> sklearn.cluster <span class="im">import</span> SpectralClustering</span>
<span id="cb1-27"><a href="#cb1-27" tabindex="-1"></a><span class="im">from</span> sklearn.cluster <span class="im">import</span> AffinityPropagation</span>
<span id="cb1-28"><a href="#cb1-28" tabindex="-1"></a></span>
<span id="cb1-29"><a href="#cb1-29" tabindex="-1"></a></span>
<span id="cb1-30"><a href="#cb1-30" tabindex="-1"></a><span class="co">##Plotting Results###</span></span>
<span id="cb1-31"><a href="#cb1-31" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-32"><a href="#cb1-32" tabindex="-1"></a><span class="im">from</span> matplotlib.ticker <span class="im">import</span> LinearLocator, FormatStrFormatter</span>
<span id="cb1-33"><a href="#cb1-33" tabindex="-1"></a><span class="im">from</span> mpl_toolkits <span class="im">import</span> mplot3d</span>
<span id="cb1-34"><a href="#cb1-34" tabindex="-1"></a><span class="im">from</span> matplotlib.pyplot <span class="im">import</span> subplots, axes, scatter, xticks</span>
<span id="cb1-35"><a href="#cb1-35" tabindex="-1"></a><span class="im">from</span> matplotlib.colors <span class="im">import</span> ListedColormap</span>
<span id="cb1-36"><a href="#cb1-36" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-37"><a href="#cb1-37" tabindex="-1"></a><span class="im">from</span> sklearn.metrics <span class="im">import</span> confusion_matrix</span>
<span id="cb1-38"><a href="#cb1-38" tabindex="-1"></a><span class="im">from</span> sklearn.metrics.cluster <span class="im">import</span> adjusted_rand_score</span>
<span id="cb1-39"><a href="#cb1-39" tabindex="-1"></a><span class="im">from</span> sklearn <span class="im">import</span> metrics</span>
<span id="cb1-40"><a href="#cb1-40" tabindex="-1"></a><span class="im">from</span> sklearn.metrics <span class="im">import</span> matthews_corrcoef</span>
<span id="cb1-41"><a href="#cb1-41" tabindex="-1"></a></span>
<span id="cb1-42"><a href="#cb1-42" tabindex="-1"></a></span>
<span id="cb1-43"><a href="#cb1-43" tabindex="-1"></a><span class="co">##Model Optimisation###</span></span>
<span id="cb1-44"><a href="#cb1-44" tabindex="-1"></a><span class="im">from</span> sklearn.decomposition <span class="im">import</span> PCA</span>
<span id="cb1-45"><a href="#cb1-45" tabindex="-1"></a><span class="im">from</span> sklearn.model_selection <span class="im">import</span> GridSearchCV</span>
<span id="cb1-46"><a href="#cb1-46" tabindex="-1"></a><span class="im">from</span> sklearn.metrics <span class="im">import</span> make_scorer</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>n<span class="op">=</span><span class="dv">2</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>RANDOM_SEED<span class="op">=</span><span class="dv">42</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co">#    'KNeighbours': KNeighborsClassifier(2),</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a> <span class="co">#   'Guassian':GaussianMixture(n_components=n),</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a> <span class="co">#   'KMeans': KMeans(n_clusters=n) ,</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>classifiers <span class="op">=</span> {</span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>    <span class="st">'SVC'</span>:SVC(kernel<span class="op">=</span><span class="st">"linear"</span>, C<span class="op">=</span><span class="fl">0.025</span>),</span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a>    <span class="st">'SVC2'</span>: SVC(gamma<span class="op">=</span><span class="dv">2</span>, C<span class="op">=</span><span class="dv">1</span>),</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>    <span class="st">'DecisionTree'</span>: DecisionTreeClassifier(max_depth<span class="op">=</span><span class="dv">5</span>),</span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a>    <span class="st">'RFC'</span>: RandomForestClassifier(max_depth<span class="op">=</span><span class="dv">5</span>, n_estimators<span class="op">=</span><span class="dv">10</span>, max_features<span class="op">=</span><span class="dv">1</span>),</span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a>    <span class="st">'MLPC'</span>: MLPClassifier(alpha<span class="op">=</span><span class="dv">1</span>, max_iter<span class="op">=</span><span class="dv">1000</span>, random_state<span class="op">=</span><span class="dv">42</span>),</span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a>    <span class="st">'ADABoost'</span>:AdaBoostClassifier(),</span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a>    <span class="st">'GaussianNB'</span>: GaussianNB(),</span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a>    <span class="st">'QDA'</span>:QuadraticDiscriminantAnalysis(),</span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a>}</span></code></pre>
</div>
</section><section id="generating-the-encoded-dataset"><h2 class="section-heading">Generating the Encoded Dataset<a class="anchor" aria-label="anchor" href="#generating-the-encoded-dataset"></a>
</h2>
<hr class="half-width">
<p>Here we input our fasta file and split the entries into VH and VL
sequences. We put each set of sequecnes through the Propythia encoder a
dataframe of numerical information for both VH and VL sequences. There
are 4000 records in the fasta file representing 2000 paired antibodies:
1000 human and 1000 mouse.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="kw">def</span> Get_Kidera_Encoded_Sequences(fasta):</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>    VH_sequences <span class="op">=</span> []</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>    VL_sequences <span class="op">=</span> []</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>    Kidera_Factors <span class="op">=</span> {<span class="st">"A"</span>: [<span class="op">-</span><span class="fl">1.56</span>,<span class="op">-</span><span class="fl">1.67</span>,<span class="op">-</span><span class="fl">0.97</span>,<span class="op">-</span><span class="fl">0.27</span>,<span class="op">-</span><span class="fl">0.93</span>,<span class="op">-</span><span class="fl">0.78</span>,<span class="op">-</span><span class="fl">0.20</span>,<span class="op">-</span><span class="fl">0.08</span>,<span class="fl">0.21</span>,<span class="op">-</span><span class="fl">0.48</span>], <span class="st">"R"</span>: [<span class="fl">0.22</span>,<span class="fl">1.27</span>,<span class="fl">1.37</span>,<span class="fl">1.87</span>,<span class="op">-</span><span class="fl">1.70</span>,<span class="fl">0.46</span>,<span class="fl">0.92</span>,<span class="op">-</span><span class="fl">0.39</span>,<span class="fl">0.23</span>,<span class="fl">0.93</span>], <span class="st">"N"</span>: [<span class="fl">1.14</span>,<span class="op">-</span><span class="fl">0.07</span>,<span class="op">-</span><span class="fl">0.12</span>,<span class="fl">0.81</span>,<span class="fl">0.18</span>,<span class="fl">0.37</span>,<span class="op">-</span><span class="fl">0.09</span>,<span class="fl">1.23</span>,<span class="fl">1.10</span>,<span class="op">-</span><span class="fl">1.73</span>], <span class="st">"D"</span>: [<span class="fl">0.58</span>,<span class="op">-</span><span class="fl">0.22</span>,<span class="op">-</span><span class="fl">1.58</span>,<span class="fl">0.81</span>,<span class="op">-</span><span class="fl">0.92</span>,<span class="fl">0.15</span>,<span class="op">-</span><span class="fl">1.52</span>,<span class="fl">0.47</span>,<span class="fl">0.76</span>,<span class="fl">0.70</span>], <span class="st">"C"</span>: [<span class="fl">0.12</span>,<span class="op">-</span><span class="fl">0.89</span>,<span class="fl">0.45</span>,<span class="op">-</span><span class="fl">1.05</span>,<span class="op">-</span><span class="fl">0.71</span>,<span class="fl">2.41</span>,<span class="fl">1.52</span>,<span class="op">-</span><span class="fl">0.69</span>,<span class="fl">1.13</span>,<span class="fl">1.10</span>], <span class="st">"Q"</span>: [<span class="op">-</span><span class="fl">0.47</span>,<span class="fl">0.24</span>,<span class="fl">0.07</span>,<span class="fl">1.10</span>,<span class="fl">1.10</span>,<span class="fl">0.59</span>,<span class="fl">0.84</span>,<span class="op">-</span><span class="fl">0.71</span>,<span class="op">-</span><span class="fl">0.03</span>,<span class="op">-</span><span class="fl">2.33</span>], <span class="st">"E"</span>: [<span class="op">-</span><span class="fl">1.45</span>,<span class="fl">0.19</span>,<span class="op">-</span><span class="fl">1.61</span>,<span class="fl">1.17</span>,<span class="op">-</span><span class="fl">1.31</span>,<span class="fl">0.40</span>,<span class="fl">0.04</span>,<span class="fl">0.38</span>,<span class="op">-</span><span class="fl">0.35</span>,<span class="op">-</span><span class="fl">0.12</span>], <span class="st">"G"</span>: [<span class="fl">1.46</span>,<span class="op">-</span><span class="fl">1.96</span>,<span class="op">-</span><span class="fl">0.23</span>,<span class="op">-</span><span class="fl">0.16</span>,<span class="fl">0.10</span>,<span class="op">-</span><span class="fl">0.11</span>,<span class="fl">1.32</span>,<span class="fl">2.36</span>,<span class="op">-</span><span class="fl">1.66</span>,<span class="fl">0.46</span>], <span class="st">"H"</span>: [<span class="op">-</span><span class="fl">0.41</span>,<span class="fl">0.52</span>,<span class="op">-</span><span class="fl">0.28</span>,<span class="fl">0.28</span>,<span class="fl">1.61</span>,<span class="fl">1.01</span>,<span class="op">-</span><span class="fl">1.85</span>,<span class="fl">0.47</span>,<span class="fl">1.13</span>,<span class="fl">1.63</span>], <span class="st">"I"</span>: [<span class="op">-</span><span class="fl">0.73</span>,<span class="op">-</span><span class="fl">0.16</span>,<span class="fl">1.79</span>,<span class="op">-</span><span class="fl">0.77</span>,<span class="op">-</span><span class="fl">0.54</span>,<span class="fl">0.03</span>,<span class="op">-</span><span class="fl">0.83</span>,<span class="fl">0.51</span>,<span class="fl">0.66</span>,<span class="op">-</span><span class="fl">1.78</span>], <span class="st">"L"</span>: [<span class="op">-</span><span class="fl">1.04</span>,<span class="fl">0.00</span>,<span class="op">-</span><span class="fl">0.24</span>,<span class="op">-</span><span class="fl">1.10</span>,<span class="op">-</span><span class="fl">0.55</span>,<span class="op">-</span><span class="fl">2.05</span>,<span class="fl">0.96</span>,<span class="op">-</span><span class="fl">0.76</span>,<span class="fl">0.45</span>,<span class="fl">0.93</span>], <span class="st">"K"</span>: [<span class="op">-</span><span class="fl">0.34</span>,<span class="fl">0.82</span>,<span class="op">-</span><span class="fl">0.23</span>,<span class="fl">1.70</span>,<span class="fl">1.54</span>,<span class="op">-</span><span class="fl">1.62</span>,<span class="fl">1.15</span>,<span class="op">-</span><span class="fl">0.08</span>,<span class="op">-</span><span class="fl">0.48</span>,<span class="fl">0.60</span>], <span class="st">"M"</span>: [<span class="op">-</span><span class="fl">1.40</span>,<span class="fl">0.18</span>,<span class="op">-</span><span class="fl">0.42</span>,<span class="op">-</span><span class="fl">0.73</span>,<span class="fl">2.00</span>,<span class="fl">1.52</span>,<span class="fl">0.26</span>,<span class="fl">0.11</span>,<span class="op">-</span><span class="fl">1.27</span>,<span class="fl">0.27</span>], <span class="st">"F"</span>: [<span class="op">-</span><span class="fl">0.21</span>,<span class="fl">0.98</span>,<span class="op">-</span><span class="fl">0.36</span>,<span class="op">-</span><span class="fl">1.43</span>,<span class="fl">0.22</span>,<span class="op">-</span><span class="fl">0.81</span>,<span class="fl">0.67</span>,<span class="fl">1.10</span>,<span class="fl">1.71</span>,<span class="op">-</span><span class="fl">0.44</span>], <span class="st">"P"</span>: [<span class="fl">2.06</span>,<span class="op">-</span><span class="fl">0.33</span>,<span class="op">-</span><span class="fl">1.15</span>,<span class="op">-</span><span class="fl">0.75</span>,<span class="fl">0.88</span>,<span class="op">-</span><span class="fl">0.45</span>,<span class="fl">0.30</span>,<span class="op">-</span><span class="fl">2.30</span>,<span class="fl">0.74</span>,<span class="op">-</span><span class="fl">0.28</span>], <span class="st">"S"</span>: [<span class="fl">0.81</span>,<span class="op">-</span><span class="fl">1.08</span>,<span class="fl">0.16</span>,<span class="fl">0.42</span>,<span class="op">-</span><span class="fl">0.21</span>,<span class="op">-</span><span class="fl">0.43</span>,<span class="op">-</span><span class="fl">1.89</span>,<span class="op">-</span><span class="fl">1.15</span>,<span class="op">-</span><span class="fl">0.97</span>,<span class="op">-</span><span class="fl">0.23</span>], <span class="st">"T"</span>: [<span class="fl">0.26</span>,<span class="op">-</span><span class="fl">0.70</span>,<span class="fl">1.21</span>,<span class="fl">0.63</span>,<span class="op">-</span><span class="fl">0.10</span>,<span class="fl">0.21</span>,<span class="fl">0.24</span>,<span class="op">-</span><span class="fl">1.15</span>,<span class="op">-</span><span class="fl">0.56</span>,<span class="fl">0.19</span>], <span class="st">"W"</span>: [<span class="fl">0.30</span>,<span class="fl">2.10</span>,<span class="op">-</span><span class="fl">0.72</span>,<span class="op">-</span><span class="fl">1.57</span>,<span class="op">-</span><span class="fl">1.16</span>,<span class="fl">0.57</span>,<span class="op">-</span><span class="fl">0.48</span>,<span class="op">-</span><span class="fl">0.40</span>,<span class="op">-</span><span class="fl">2.30</span>,<span class="op">-</span><span class="fl">0.60</span>], <span class="st">"Y"</span>: [<span class="fl">1.38</span>,<span class="fl">1.48</span>,<span class="fl">0.80</span>,<span class="op">-</span><span class="fl">0.56</span>,<span class="op">-</span><span class="fl">0.00</span>,<span class="op">-</span><span class="fl">0.68</span>,<span class="op">-</span><span class="fl">0.31</span>,<span class="fl">1.03</span>,<span class="op">-</span><span class="fl">0.05</span>,<span class="fl">0.53</span>], <span class="st">"V"</span>: [<span class="op">-</span><span class="fl">0.74</span>,<span class="op">-</span><span class="fl">0.71</span>,<span class="fl">2.04</span>,<span class="op">-</span><span class="fl">0.40</span>,<span class="fl">0.50</span>,<span class="op">-</span><span class="fl">0.81</span>,<span class="op">-</span><span class="fl">1.07</span>,<span class="fl">0.06</span>,<span class="op">-</span><span class="fl">0.46</span>,<span class="fl">0.65</span>], <span class="st">"X"</span>: [<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>]}</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>    <span class="cf">with</span> <span class="bu">open</span>(fasta, <span class="st">"r"</span>) <span class="im">as</span> f:</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>        <span class="cf">for</span> line <span class="kw">in</span> f:</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>            <span class="cf">if</span> line[<span class="dv">0</span>] <span class="op">==</span> <span class="st">"&gt;"</span>:</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>                <span class="cf">if</span> <span class="st">"_VH"</span> <span class="kw">in</span> line:</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>                    sequence_to_add <span class="op">=</span> f.readline().strip()</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>                    VH_sequences.append(sequence_to_add)</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>                <span class="cf">elif</span> <span class="st">"_VL"</span> <span class="kw">in</span> line:</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>                    sequence_to_add <span class="op">=</span> f.readline().strip()</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>                    VL_sequences.append(sequence_to_add)</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">len</span>(VH_sequences) <span class="op">==</span> <span class="bu">len</span>(VL_sequences):</span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a>        VH_encodings <span class="op">=</span> []</span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a>        VL_encodings <span class="op">=</span> []</span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(VH_sequences)):</span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a>            VH_kidera <span class="op">=</span> []</span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a>            <span class="cf">for</span> res <span class="kw">in</span> VH_sequences[i]:</span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a>                VH_kidera <span class="op">=</span> VH_kidera <span class="op">+</span> Kidera_Factors[res]</span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a>            VH_encodings.append(VH_kidera)</span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(VL_sequences)):</span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a>            VL_kidera <span class="op">=</span> []</span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a>            <span class="cf">for</span> res <span class="kw">in</span> VL_sequences[i]:</span>
<span id="cb3-26"><a href="#cb3-26" tabindex="-1"></a>                VL_kidera <span class="op">=</span> VL_kidera <span class="op">+</span> Kidera_Factors[res]</span>
<span id="cb3-27"><a href="#cb3-27" tabindex="-1"></a>            VL_encodings.append(VL_kidera)</span>
<span id="cb3-28"><a href="#cb3-28" tabindex="-1"></a>    VH_dataframe <span class="op">=</span> pd.DataFrame(VH_encodings)</span>
<span id="cb3-29"><a href="#cb3-29" tabindex="-1"></a>    VL_dataframe <span class="op">=</span> pd.DataFrame(VL_encodings)</span>
<span id="cb3-30"><a href="#cb3-30" tabindex="-1"></a>    <span class="co"># Now we join these two dataframes together so that each sample now has information about its VH and VL sequence.</span></span>
<span id="cb3-31"><a href="#cb3-31" tabindex="-1"></a>    VH_dataframe_suffix <span class="op">=</span> VH_dataframe.add_suffix(<span class="st">'_VH'</span>)</span>
<span id="cb3-32"><a href="#cb3-32" tabindex="-1"></a>    VL_dataframe_suffix <span class="op">=</span> VL_dataframe.add_suffix(<span class="st">'_VL'</span>)</span>
<span id="cb3-33"><a href="#cb3-33" tabindex="-1"></a>    joined_dataframe_VH_VL <span class="op">=</span>  VH_dataframe_suffix.join(VL_dataframe_suffix)</span>
<span id="cb3-34"><a href="#cb3-34" tabindex="-1"></a>    joined_dataframe_VH_VL_cleaned <span class="op">=</span> joined_dataframe_VH_VL.dropna(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb3-35"><a href="#cb3-35" tabindex="-1"></a>    <span class="cf">return</span>(joined_dataframe_VH_VL_cleaned)</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>input_fasta <span class="op">=</span> <span class="st">'data/HumanMouseOAS_VH_VL_paired_data.faa'</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>joined_dataframe_VH_VL <span class="op">=</span> Get_Kidera_Encoded_Sequences(input_fasta)</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="bu">print</span>(joined_dataframe_VH_VL)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>      0_VH  1_VH  2_VH  3_VH  4_VH  ...  725_VL  726_VL  727_VL  728_VL  729_VL
0    -0.47  0.24  0.07  1.10  1.10  ...   -0.78   -0.20   -0.08    0.21   -0.48
1    -0.47  0.24  0.07  1.10  1.10  ...   -2.05    0.96   -0.76    0.45    0.93
2    -1.45  0.19 -1.61  1.17 -1.31  ...    0.21    0.24   -1.15   -0.56    0.19
3    -1.45  0.19 -1.61  1.17 -1.31  ...   -0.78   -0.20   -0.08    0.21   -0.48
4    -0.47  0.24  0.07  1.10  1.10  ...   -0.11    1.32    2.36   -1.66    0.46
...    ...   ...   ...   ...   ...  ...     ...     ...     ...     ...     ...
1995 -0.47  0.24  0.07  1.10  1.10  ...   -0.11    1.32    2.36   -1.66    0.46
1996 -1.45  0.19 -1.61  1.17 -1.31  ...   -0.43   -1.89   -1.15   -0.97   -0.23
1997 -0.47  0.24  0.07  1.10  1.10  ...   -2.05    0.96   -0.76    0.45    0.93
1998 -0.47  0.24  0.07  1.10  1.10  ...   -0.78   -0.20   -0.08    0.21   -0.48
1999 -1.45  0.19 -1.61  1.17 -1.31  ...   -0.43   -1.89   -1.15   -0.97   -0.23

[2000 rows x 1780 columns]</code></pre>
</div>
<p>Now we have our encodings, we need to prepare our labels. As our
input was ordered 1000 Human antibodies and 1000 Mouse antibodies we can
simply make a list showing only these:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co">#Prepare training data and labels</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>labels1 <span class="op">=</span> <span class="dv">1000</span><span class="op">*</span>[<span class="dv">1</span>] <span class="co">##Human antibodies will be class 1</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>labels2 <span class="op">=</span> <span class="dv">1000</span><span class="op">*</span>[<span class="dv">0</span>] <span class="co">## Mouse antibodies will be class 0</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>labels <span class="op">=</span> labels1<span class="op">+</span>labels2</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>y<span class="op">=</span>labels</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(y))</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="fl">2000</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co">##Mouse ==1, Human == 0</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>dataset <span class="op">=</span> joined_dataframe_VH_VL</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>dataset<span class="op">=</span>dataset.loc[:, dataset.columns <span class="op">!=</span> <span class="st">'Unnamed: 0'</span>]</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="bu">print</span>(dataset.shape) <span class="co">##Just to check that you have an equal number of labels to the number of samples</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>(2000, 1780)</code></pre>
</div>
<p>Now that we have our datasets, we may not split them into training
datasets for fitting our classifiers to and training datasets to verify
their effectiveness as predictors. Usually 70/30 or 80/20 split is
necessary.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>X_train, X_test, y_train, y_test <span class="op">=</span> train_test_split(dataset, y, test_size<span class="op">=</span><span class="fl">.3</span>, random_state<span class="op">=</span>RANDOM_SEED, shuffle<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>num_rows, num_cols <span class="op">=</span> dataset.shape</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Training set size: "</span>, X_train.shape, <span class="st">"       Test set size: "</span>, X_test.shape)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Training set size:  (1400, 1780)        Test set size:  (600, 1780)</code></pre>
</div>
</section><section id="separating-our-data-with-machine-learning-classifiers"><h2 class="section-heading">Separating our data with Machine Learning Classifiers<a class="anchor" aria-label="anchor" href="#separating-our-data-with-machine-learning-classifiers"></a>
</h2>
<hr class="half-width">
<p>Here is our list of classifiers that we will loop through to see
which is the best at clustering our dataset:</p>
<p>Now loop over our classifiers and use the test and train datasets to
generate a score to validate the classifiers. I have chosen Matthews
Correlation Coefficient (MCC) which is a metric less prone to bias by
taking into account false predictions, as well as true predictions. This
metric lies on a score between -1 (inverse prediction) and 1 (perfect
prediction) with 0 being coin toss likelihood. We then plot our results
as a confusion matrix which demonstrates the predictive power of our
classifiers. The Confusion matrix shows the raw number of records that
have been assigned to each category in a 2x2 matrix and is given as
such:</p>
<table class="table">
<thead><tr class="header">
<th></th>
<th>Predicted Class = 0</th>
<th>Predicted Class = 1</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Actual Class = 0</td>
<td>True Negative</td>
<td>False Positive</td>
</tr>
<tr class="even">
<td>Actual Class = 1</td>
<td>False Negative</td>
<td>True Positive</td>
</tr>
</tbody>
</table>
<p>Ideally we want the True Negative and True Positive field to be the
most popular fields with only a few records in the false positive
fields.</p>
<p>N.B.: Results may vary between each run due to the stochastic nature
of the machine learning algorithms.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co">##Loop through each classifier, fit training data and evaluate model. Plot results as confusion matrix##</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>scores <span class="op">=</span> []</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> classifiers:</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>    clf_1 <span class="op">=</span> classifiers.get(i)</span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>    clf_1.fit(X_train,y_train)</span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a>    y_predict1 <span class="op">=</span> clf_1.predict(X_test)</span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a>    scoring <span class="op">=</span> matthews_corrcoef(y_test, y_predict1)</span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a>    scores.append(scoring) </span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a>    confusion_matrix <span class="op">=</span> metrics.confusion_matrix(y_test, y_predict1)</span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a>    ax1 <span class="op">=</span> sns.heatmap(confusion_matrix, annot<span class="op">=</span><span class="va">True</span>, cmap<span class="op">=</span><span class="st">'summer'</span>)</span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a>    title <span class="op">=</span> <span class="bu">str</span>(i)</span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a>    ax1.set_title(title)<span class="op">;</span></span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a>    ax1.set_xlabel(<span class="st">'</span><span class="ch">\n</span><span class="st">Predicted Values'</span>)</span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a>    ax1.set_ylabel(<span class="st">'Actual Values '</span>)<span class="op">;</span></span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a>    plt.show()</span></code></pre>
</div>
<p><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" class="figure"><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-7-2.png" width="672" style="display: block; margin: auto;" class="figure"><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-7-3.png" width="672" style="display: block; margin: auto;" class="figure"><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-7-4.png" width="672" style="display: block; margin: auto;" class="figure"><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-7-5.png" width="672" style="display: block; margin: auto;" class="figure"><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-7-6.png" width="672" style="display: block; margin: auto;" class="figure"><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-7-7.png" width="672" style="display: block; margin: auto;" class="figure"><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-7-8.png" width="672" style="display: block; margin: auto;" class="figure"></p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co">##Plot Performance of all Models##</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">8</span>,<span class="dv">6</span>))</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>plt.suptitle(<span class="st">'Performance of Machine Learning Classifiers Against Mouse and Human Antibodies'</span>, fontsize<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>bins <span class="op">=</span> np.arange(<span class="bu">len</span>(classifiers))</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>ax.bar(np.arange(<span class="bu">len</span>(scores)), scores)</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>ax.set_ylabel(<span class="st">'Matthews Correlation Coefficient'</span>)</span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a>ax.set_xlabel(<span class="st">'Classifiers'</span>)</span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a>ax.set_xticks(bins)</span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a>ax.set_xticklabels(classifiers, rotation<span class="op">=-</span><span class="dv">80</span>)<span class="op">;</span></span></code></pre>
</div>
<figure><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-8-17.png" width="768" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>As we can see, the majority of these classifiers have been very
successful at training on the data and learning to classify the two
groups, with nearly perfect performance in the SVC model and ADABoost.
What we can assume from this is that there are highly correlated
features in the training data which the models have used for training.
However, the Kidera factor is not the only method of encoding sequences
for machine learning algorithms. In fact, for more complicated problems,
usually deeper methods of encodings are required.</p>
<div class="section level3">
<h3 id="using-amino-acid-compositions-to-encode-preotein-sequences">Using Amino Acid Compositions to Encode Preotein Sequences<a class="anchor" aria-label="anchor" href="#using-amino-acid-compositions-to-encode-preotein-sequences"></a>
</h3>
<p>Amino acid compositions are more sequence based statistics than the
physiochemical properties evaluated in the Kidera factors. These
generate much denser and complex encodings which we will compare with
the Kidera factor method.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co">###Propythia Command to get encodings###</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="kw">def</span> get_descriptors(protein):</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>    <span class="co">##This will retrieve a selection of encodings that are not dependent on the sequence length##</span></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>    test<span class="op">=</span> protein.adaptable([<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">6</span>,<span class="dv">7</span>,<span class="dv">8</span>,<span class="dv">9</span>,<span class="dv">10</span>,<span class="dv">11</span>,<span class="dv">12</span>,<span class="dv">13</span>,<span class="dv">14</span>,<span class="dv">17</span>,<span class="dv">18</span>,<span class="dv">19</span>,<span class="dv">20</span>,<span class="dv">21</span>])</span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a>    <span class="cf">return</span>(test)</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="kw">def</span> Get_dataset(fasta):</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>    VH_sequences <span class="op">=</span> []</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>    VL_sequences <span class="op">=</span> []</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a>    <span class="cf">with</span> <span class="bu">open</span>(fasta, <span class="st">"r"</span>) <span class="im">as</span> f:</span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a>        <span class="cf">for</span> line <span class="kw">in</span> f:</span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a>            <span class="cf">if</span> line[<span class="dv">0</span>] <span class="op">==</span> <span class="st">"&gt;"</span>:</span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a>                <span class="cf">if</span> <span class="st">"_VH"</span> <span class="kw">in</span> line:</span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a>                    sequence_to_add <span class="op">=</span> f.readline().strip()</span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a>                    VH_sequences.append(sequence_to_add)</span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a>                <span class="cf">elif</span> <span class="st">"_VL"</span> <span class="kw">in</span> line:</span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a>                    sequence_to_add <span class="op">=</span> f.readline().strip()</span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a>                    VL_sequences.append(sequence_to_add)</span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a></span>
<span id="cb15-14"><a href="#cb15-14" tabindex="-1"></a>    <span class="bu">print</span>(<span class="bu">len</span>(VH_sequences),<span class="bu">len</span>(VL_sequences))</span>
<span id="cb15-15"><a href="#cb15-15" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">len</span>(VH_sequences) <span class="op">==</span> <span class="bu">len</span>(VL_sequences):</span>
<span id="cb15-16"><a href="#cb15-16" tabindex="-1"></a>        VH_dataframe <span class="op">=</span> pd.DataFrame()</span>
<span id="cb15-17"><a href="#cb15-17" tabindex="-1"></a>        VL_dataframe <span class="op">=</span> pd.DataFrame()</span>
<span id="cb15-18"><a href="#cb15-18" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(VH_sequences)):</span>
<span id="cb15-19"><a href="#cb15-19" tabindex="-1"></a>            ps_string<span class="op">=</span>sequence.read_protein_sequence(VH_sequences[i])</span>
<span id="cb15-20"><a href="#cb15-20" tabindex="-1"></a>            protein <span class="op">=</span> Descriptor(ps_string)</span>
<span id="cb15-21"><a href="#cb15-21" tabindex="-1"></a>            descriptors <span class="op">=</span> get_descriptors(protein)</span>
<span id="cb15-22"><a href="#cb15-22" tabindex="-1"></a>            <span class="co">#VH_dataframe.loc[len(VH_dataframe)] = descriptors</span></span>
<span id="cb15-23"><a href="#cb15-23" tabindex="-1"></a>            VH_dataframe <span class="op">=</span> VH_dataframe._append(descriptors, ignore_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb15-24"><a href="#cb15-24" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"VH_data"</span>, VH_dataframe.shape)</span>
<span id="cb15-25"><a href="#cb15-25" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(VL_sequences)):</span>
<span id="cb15-26"><a href="#cb15-26" tabindex="-1"></a>            ps_string<span class="op">=</span>sequence.read_protein_sequence(VL_sequences[i])</span>
<span id="cb15-27"><a href="#cb15-27" tabindex="-1"></a>            protein <span class="op">=</span> Descriptor(ps_string)</span>
<span id="cb15-28"><a href="#cb15-28" tabindex="-1"></a>            descriptors <span class="op">=</span> get_descriptors(protein)</span>
<span id="cb15-29"><a href="#cb15-29" tabindex="-1"></a>            <span class="co">#VL_dataframe.loc[len(VL_dataframe)] = descriptors</span></span>
<span id="cb15-30"><a href="#cb15-30" tabindex="-1"></a>            VL_dataframe <span class="op">=</span> VL_dataframe._append(descriptors, ignore_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb15-31"><a href="#cb15-31" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"VL_data"</span>, VL_dataframe.shape)</span>
<span id="cb15-32"><a href="#cb15-32" tabindex="-1"></a>    <span class="co"># Now we join these two dataframes together so that each sample now has information about its VH and VL sequence.</span></span>
<span id="cb15-33"><a href="#cb15-33" tabindex="-1"></a>    VH_dataframe_suffix <span class="op">=</span> VH_dataframe.add_suffix(<span class="st">'_VH'</span>)</span>
<span id="cb15-34"><a href="#cb15-34" tabindex="-1"></a>    VL_dataframe_suffix <span class="op">=</span> VL_dataframe.add_suffix(<span class="st">'_VL'</span>)</span>
<span id="cb15-35"><a href="#cb15-35" tabindex="-1"></a>    joined_dataframe_VH_VL <span class="op">=</span>  VH_dataframe_suffix.join(VL_dataframe_suffix)</span>
<span id="cb15-36"><a href="#cb15-36" tabindex="-1"></a>    <span class="cf">return</span>(joined_dataframe_VH_VL)</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="co">### Input Fasta and Run Dataset###</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>input_fasta <span class="op">=</span> <span class="st">'data/HumanMouseOAS_VH_VL_paired_data.faa'</span></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a><span class="co">#joined_dataframe_VH_VL = Get_dataset(input_fasta)</span></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a></span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a><span class="co">#Optionally save dataframe as a CSV to simply reload it in future</span></span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a><span class="co">#joined_dataframe_VH_VL.to_csv('./HumanMouseOAS_VH_VL_paired_data.faa_Full_descriptors')#</span></span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a>joined_dataframe_VH_VL <span class="op">=</span> read_csv(<span class="st">'data/HumanMouseOAS_VH_VL_paired_data.faa_Full_descriptors.csv'</span>, header <span class="op">=</span> <span class="dv">0</span>)</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="do-it-yourself">Do it yourself<a class="anchor" aria-label="anchor" href="#do-it-yourself"></a>
</h3>
<ul>
<li><p>These encodings used with Propythia were selected to reduce the
time taken to run. Retry the encoding step and experiment with the
protein.adaptable([3,4,5,6,7,8,9,10,11,12,13,14,17,18,19,20,21])
array.</p></li>
<li><p>NB Propythia accepts numbers 0-40 however we avoid 1, 2 and 37 as
these produce outputs of differing length</p></li>
</ul>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>RANDOM_SEED <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="co">#Prepare training data and labels</span></span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a>labels1 <span class="op">=</span> <span class="dv">1000</span><span class="op">*</span>[<span class="dv">1</span>] <span class="co">##Human antibodies will be class 1</span></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a>labels2 <span class="op">=</span> <span class="dv">1000</span><span class="op">*</span>[<span class="dv">0</span>] <span class="co">## Mouse antibodies will be class 0</span></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a>labels <span class="op">=</span> labels1<span class="op">+</span>labels2</span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a>y<span class="op">=</span>labels</span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(y))</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="fl">2000</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="co">##Mouse ==1, Human == 0</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>dataset <span class="op">=</span> joined_dataframe_VH_VL</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>dataset<span class="op">=</span>dataset.loc[:, dataset.columns <span class="op">!=</span> <span class="st">'Unnamed: 0'</span>]</span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a><span class="bu">print</span>(dataset.shape) <span class="co">##Just to check that you have an equal number of labels to the number of samples</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>(2000, 890)</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>X_train, X_test, y_train, y_test <span class="op">=</span> train_test_split(dataset, y, test_size<span class="op">=</span><span class="fl">.3</span>, random_state<span class="op">=</span>RANDOM_SEED, shuffle<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>num_rows, num_cols <span class="op">=</span> dataset.shape</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Training set size: "</span>, X_train.shape, <span class="st">"       Test set size: "</span>, X_test.shape)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Training set size:  (1400, 890)        Test set size:  (600, 890)</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a>n<span class="op">=</span><span class="dv">2</span></span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>RANDOM_SEED<span class="op">=</span><span class="dv">42</span></span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a></span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a></span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a></span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a><span class="co">#    'KNeighbours': KNeighborsClassifier(2),</span></span>
<span id="cb23-7"><a href="#cb23-7" tabindex="-1"></a> <span class="co">#   'Guassian':GaussianMixture(n_components=n),</span></span>
<span id="cb23-8"><a href="#cb23-8" tabindex="-1"></a> <span class="co">#   'KMeans': KMeans(n_clusters=n) ,</span></span>
<span id="cb23-9"><a href="#cb23-9" tabindex="-1"></a></span>
<span id="cb23-10"><a href="#cb23-10" tabindex="-1"></a>classifiers <span class="op">=</span> {</span>
<span id="cb23-11"><a href="#cb23-11" tabindex="-1"></a>    <span class="st">'SVC'</span>:SVC(kernel<span class="op">=</span><span class="st">"linear"</span>, C<span class="op">=</span><span class="fl">0.025</span>),</span>
<span id="cb23-12"><a href="#cb23-12" tabindex="-1"></a>    <span class="st">'SVC2'</span>: SVC(gamma<span class="op">=</span><span class="dv">2</span>, C<span class="op">=</span><span class="dv">1</span>),</span>
<span id="cb23-13"><a href="#cb23-13" tabindex="-1"></a>    <span class="st">'DecisionTree'</span>: DecisionTreeClassifier(max_depth<span class="op">=</span><span class="dv">5</span>),</span>
<span id="cb23-14"><a href="#cb23-14" tabindex="-1"></a>    <span class="st">'RFC'</span>: RandomForestClassifier(max_depth<span class="op">=</span><span class="dv">5</span>, n_estimators<span class="op">=</span><span class="dv">10</span>, max_features<span class="op">=</span><span class="dv">1</span>),</span>
<span id="cb23-15"><a href="#cb23-15" tabindex="-1"></a>    <span class="st">'MLPC'</span>: MLPClassifier(alpha<span class="op">=</span><span class="dv">1</span>, max_iter<span class="op">=</span><span class="dv">1000</span>),</span>
<span id="cb23-16"><a href="#cb23-16" tabindex="-1"></a>    <span class="st">'ADABoost'</span>:AdaBoostClassifier(),</span>
<span id="cb23-17"><a href="#cb23-17" tabindex="-1"></a>    <span class="st">'GaussianNB'</span>: GaussianNB(),</span>
<span id="cb23-18"><a href="#cb23-18" tabindex="-1"></a>    <span class="st">'QDA'</span>:QuadraticDiscriminantAnalysis(),</span>
<span id="cb23-19"><a href="#cb23-19" tabindex="-1"></a>}</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="do-it-yourself-1">Do it yourself<a class="anchor" aria-label="anchor" href="#do-it-yourself-1"></a>
</h3>
<ul>
<li>This is not an exhaustive list of classifiers. These were mostly
picked to represent all of the different kinds of models. Here you will
find a much larger list of classifiers. Try adding some to the
classifiers dictionary and see how the results differ. Additional models
that are supported in scikit learn can be found here: <a href="https://scikit-learn.org/stable/supervised_learning.html" class="external-link uri">https://scikit-learn.org/stable/supervised_learning.html</a>
</li>
</ul>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="co">##Loop through each classifier, fit training data and evaluate model. Plot results as confusion matrix##</span></span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a>scores <span class="op">=</span> []</span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> classifiers:</span>
<span id="cb24-5"><a href="#cb24-5" tabindex="-1"></a>    clf_1 <span class="op">=</span> classifiers.get(i)</span>
<span id="cb24-6"><a href="#cb24-6" tabindex="-1"></a>    clf_1.fit(X_train,y_train)</span>
<span id="cb24-7"><a href="#cb24-7" tabindex="-1"></a>    y_predict1 <span class="op">=</span> clf_1.predict(X_test)</span>
<span id="cb24-8"><a href="#cb24-8" tabindex="-1"></a>    scoring <span class="op">=</span> matthews_corrcoef(y_test, y_predict1)</span>
<span id="cb24-9"><a href="#cb24-9" tabindex="-1"></a>    scores.append(scoring) </span>
<span id="cb24-10"><a href="#cb24-10" tabindex="-1"></a>    confusion_matrix <span class="op">=</span> metrics.confusion_matrix(y_test, y_predict1)</span>
<span id="cb24-11"><a href="#cb24-11" tabindex="-1"></a>    ax1 <span class="op">=</span> sns.heatmap(confusion_matrix, annot<span class="op">=</span><span class="va">True</span>, cmap<span class="op">=</span><span class="st">'summer'</span>)</span>
<span id="cb24-12"><a href="#cb24-12" tabindex="-1"></a>    title <span class="op">=</span> <span class="bu">str</span>(i)</span>
<span id="cb24-13"><a href="#cb24-13" tabindex="-1"></a>    ax1.set_title(title)<span class="op">;</span></span>
<span id="cb24-14"><a href="#cb24-14" tabindex="-1"></a>    ax1.set_xlabel(<span class="st">'</span><span class="ch">\n</span><span class="st">Predicted Values'</span>)</span>
<span id="cb24-15"><a href="#cb24-15" tabindex="-1"></a>    ax1.set_ylabel(<span class="st">'Actual Values '</span>)<span class="op">;</span></span>
<span id="cb24-16"><a href="#cb24-16" tabindex="-1"></a>    plt.show()</span></code></pre>
</div>
<p><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-15-19.png" width="672" style="display: block; margin: auto;" class="figure"><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-15-20.png" width="672" style="display: block; margin: auto;" class="figure"><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-15-21.png" width="672" style="display: block; margin: auto;" class="figure"><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-15-22.png" width="672" style="display: block; margin: auto;" class="figure"><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-15-23.png" width="672" style="display: block; margin: auto;" class="figure"><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-15-24.png" width="672" style="display: block; margin: auto;" class="figure"><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-15-25.png" width="672" style="display: block; margin: auto;" class="figure"><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-15-26.png" width="672" style="display: block; margin: auto;" class="figure"></p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="co">##Plot Performance of all Models##</span></span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">8</span>,<span class="dv">6</span>))</span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a>plt.suptitle(<span class="st">'Performance of Machine Learning Classifiers Against Mouse and Human Antibodies'</span>, fontsize<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a></span>
<span id="cb25-5"><a href="#cb25-5" tabindex="-1"></a>bins <span class="op">=</span> np.arange(<span class="bu">len</span>(classifiers))</span>
<span id="cb25-6"><a href="#cb25-6" tabindex="-1"></a>ax.bar(np.arange(<span class="bu">len</span>(scores)), scores)</span>
<span id="cb25-7"><a href="#cb25-7" tabindex="-1"></a>ax.set_ylabel(<span class="st">'Matthews Correlation Coefficient'</span>)</span>
<span id="cb25-8"><a href="#cb25-8" tabindex="-1"></a>ax.set_xlabel(<span class="st">'Classifiers'</span>)</span>
<span id="cb25-9"><a href="#cb25-9" tabindex="-1"></a>ax.set_xticks(bins)</span>
<span id="cb25-10"><a href="#cb25-10" tabindex="-1"></a>ax.set_xticklabels(classifiers, rotation<span class="op">=-</span><span class="dv">80</span>)<span class="op">;</span></span></code></pre>
</div>
<figure><img src="fig/Machine_Learning_Part1-rendered-unnamed-chunk-16-35.png" width="768" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>From the above chart, we can see that the best performing predictors
are ADA_Boost, GaussinNB, DecisionTree and SVC are the best performing
classifier, wheras both Gaussian and KMeans are the worst performing
with negative MCC scores.</p>
<div id="questions" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="questions" class="callout-inner">
<h3 class="callout-title">Questions<a class="anchor" aria-label="anchor" href="#questions"></a>
</h3>
<div class="callout-content">
<ul>
<li>Which encoding method produced overall the best results? Why could
this be? Hint: think of the size of both sets of encodings</li>
<li>What could be the advantages and disadvantages of both sets of
encodings?</li>
</ul>
</div>
</div>
</div>
<div id="end-of-chapter-exercise-testing-our-classifiers-on-a-naïve-dataset" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="end-of-chapter-exercise-testing-our-classifiers-on-a-naïve-dataset" class="callout-inner">
<h3 class="callout-title">End of Chapter Exercise: Testing our
Classifiers on a Naïve Dataset<a class="anchor" aria-label="anchor" href="#end-of-chapter-exercise-testing-our-classifiers-on-a-na%C3%AFve-dataset"></a>
</h3>
<div class="callout-content">
<p>We have seen that it is possible to separate mouse and human antibody
protein sequences through their numerical encodings.</p>
<p>We can also take a totally naïve dataset that the model has not been
exposed to. This is a measure that checks for overfitting. If we see
that there is poor performance on this naïve “held back” dataset, then
it could suggest overfitting to the training data. Using 20 Human and 20
mouse paired sequences from OAS, which were not used to train our
models, it is possible to generate their encodings, and pass them
through the optimised model, in order to test it.</p>
<p>The file below has 20 human and 20 mouse sequences which are held
back from our original training data. Using the skills you have learned
so far in this notebook, encode these paired sequences and generate a
list of labels for these entries. Pass them through the trained
classifiers and evaluate their performance.</p>
<ul>
<li>Comment on which classifier performs best.</li>
<li>Think of ways in which the classifiers can be improved. These will
be expanded upon in the next Notebook.</li>
</ul>
<p><code>naive_fasta = './Naive_dataset.faa.txt'</code></p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Protein sequences must be numerically encoded to be readable by
machine learning algorithms.</li>
<li>It is necessary to test your classifier with multiple models and
multiple encoding methods to find which works the best</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/LearnToDiscover/case-study-antibodies/edit/main/README.md" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/LearnToDiscover/case-study-antibodies/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/LearnToDiscover/case-study-antibodies/" class="external-link">Source</a></p>
				<p><a href="https://github.com/LearnToDiscover/case-study-antibodies/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:team@carpentries.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/LearnToDiscover/sandpaper/tree/l2d" class="external-link">sandpaper (0.16.4)</a>, <a href="https://github.com/carpentries/pegboard/tree/a32a7836d4455f407c3cafe8ab95edc636e5e919" class="external-link">pegboard (0.7.5)</a>, and <a href="https://github.com/LearnToDiscover/varnish/tree/7ac5e27a3dfed3a59ec1e5fcb0b719d430806598" class="external-link">varnish (1.0.2)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://LearnToDiscover.github.io/case-study-antibodies/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://LearnToDiscover.github.io/case-study-antibodies/aio.html",
  "identifier": "https://LearnToDiscover.github.io/case-study-antibodies/aio.html",
  "dateCreated": "2024-03-04",
  "dateModified": "2024-04-30",
  "datePublished": "2024-04-30"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

